File: solutions/javascript/difference-of-squares/1/difference-of-squares.js
export class Squares {
  #firstN;
  #sumOfSquares = undefined;
  #squareOfSum = undefined;
  #difference = undefined;

  constructor(firstN) {
    this.#firstN = [...Array(firstN).keys()].map((e) => e + 1);
  }

  get sumOfSquares() {
    if (this.#sumOfSquares == undefined) {
      this.#sumOfSquares = this.#firstN
        .map((e) => e * e)
        .reduce((acc, cur) => acc + cur, 0);
    }
    return this.#sumOfSquares;
  }

  get squareOfSum() {
    if (this.#squareOfSum == undefined) {
      let sum = this.#firstN.reduce((acc, cur) => acc + cur, 0);
      this.#squareOfSum = sum * sum;
    }
    return this.#squareOfSum;
  }

  get difference() {
    if (this.#difference == undefined) {
      this.#difference = this.squareOfSum - this.sumOfSquares;
    }
    return this.#difference;
  }
}

File: solutions/cpp/difference-of-squares/1/difference_of_squares.cpp
#include "difference_of_squares.h"
#include <numeric>
#include <algorithm>
#include <vector>
#include <iostream>

using namespace std;

namespace difference_of_squares
{
    uint64_t square_of_sum(uint32_t n)
    {
        uint32_t sum = 0;
        for (uint32_t i = 1; i <= n; i++)
            sum += i;

        return sum * sum;
    }

    uint64_t sum_of_squares(uint32_t n)
    {
        vector<uint32_t> sequence(n);
        vector<uint32_t> squares(n);
        iota(begin(sequence), end(sequence), 1);

        transform(
            begin(sequence), end(sequence), begin(squares),
            [](auto value) -> auto { return value * value; });

        uint32_t sum = 0;
        for (uint32_t square : squares)
            sum += square;

        return sum;
    }

    uint64_t difference(uint32_t n)
    {
        return square_of_sum(n) - sum_of_squares(n);
    }

} // namespace difference_of_squares
